balboa-client/src/test/scala/com/blist/metrics/impl/queue/MetricsBufferSpec.scala

```diff
@@ -35,8 +38,185 @@ class MetricsBufferSpec extends WordSpec {
+    }
   }
 
+  trait FullBufferSetup extends BufferSetup {
+    val entId1 = "ayn"
+    val entId2 = "marc"
+    val time = 1L
+    val time2 = 1L + 2*granularity
+
+    val original =  Seq(
+      (entId1, met1, time),
+      (entId1, met1, time),
+      (entId2, met2, time),
+      (entId1, met2, time2)
+    )
+
+    val toInsert = original
+  }
+
+  "when adding, the buffer " should {
+    "add the given bufferItem if empty" in new BufferSetup {
+      val entId = "ayn"
+      val time = 1L
+
+      val toInsert = Seq(
+        (entId, met1, time)
+      )
+
+      val expected = Seq(
+        new MetricsBucket(entId, met1, testBuffer.nearestSlice(time))
+      )
+
+      insertAndVerify
+    }
+    "add the given bufferItem if not empty and the key doesn't yet exist because of a new entity ID" in new BufferSetup {
+      val entId1 = "ayn"
+      val entId2 = "marc"
+      val time = 1L
+
+      val toInsert = Seq(
+        (entId1, met1, time),
+        (entId2, met2, time)
+      )
+
+      val expected = Seq(
+        new MetricsBucket(entId1, met1, testBuffer.nearestSlice(time)),
+        new MetricsBucket(entId2, met2, testBuffer.nearestSlice(time))
+      )
+    }
+    "add the given bufferItem if not empty and the key doesn't yet exist because of a new timeslice" in new FullBufferSetup {
+      val metricsForEntId1 = new Metrics
+      metricsForEntId1.put("num_kitties", new Metric(agg, -2))
+      metricsForEntId1.put("num_kiddies", new Metric(agg, 2))
+      metricsForEntId1.put("num_kaleidoscopes", new Metric(abs, 4))
+
+      val expected = Seq(
+        new MetricsBucket(entId1, metricsForEntId1, testBuffer.nearestSlice(time)),
+        new MetricsBucket(entId2, met2, testBuffer.nearestSlice(time)),
+        new MetricsBucket(entId1, met2, testBuffer.nearestSlice(time2))
+      )
+
+      insertAndVerify
+    }
+
+    "update the given bufferItem if not empty and the key does exist and types are the same" in new FullBufferSetup {
+      val entId = "marc"
+      val time3 = 1L + granularity/4
+
+      override val toInsert = original ++ Seq(
+        (entId, met1, time3)
+      )
+
+      val expected = Seq(
+        new MetricsBucket(entId, met3, testBuffer.nearestSlice(time3))
+      )
+    }
+
+    "update buffer with multiple different metrics in different bufferItems" in new FullBufferSetup {
+      val time3 = 1L + granularity/4
+
+      override val toInsert = Seq(
+        (entId1, met6, time),
+        (entId1, met7, time),
+        (entId1, met8, time3),
+        (entId1, met9, time3)
+      )
+
+      val expected = Seq(
+        new MetricsBucket(entId1, met10, testBuffer.nearestSlice(time3))
+      )
+
+      insertAndVerify
+    }
+  }
+
+  def equivalentMetrics(metrics1:Metrics, metrics2:Metrics) : Boolean = {
+    val keys1 = metrics1.keySet().asScala
+    val keys2 = metrics2.keySet().asScala
+    keys1.equals(keys2) && keys1.filter(k => {
+      !Option(metrics1.get(k)).equals(Option(metrics1.get(k)))
```

Comment:
Marshaling into `Option`s seems odd here, but it might just be a strange intersection of Java and Scala.